c_src/membrane_opus/decoder.c

```diff
@@ -0,0 +1,74 @@
+#include "decoder.h"
+
+// Based on https://opus-codec.org/docs/opus_api-1.1.3/
+
+UNIFEX_TERM create(UnifexEnv *env, int sample_rate, int channels) {
+  State *state = unifex_alloc_state(env);
+  int error = 0;
+  state->decoder = opus_decoder_create(sample_rate, channels, &error);
+  state->channels = channels;
+  state->sample_rate = sample_rate;
+
+  if (error != OPUS_OK) {
+    unifex_release_state(env, state);
+    return create_result_error(env, (char *)opus_strerror(error));
+  }
+
+  UNIFEX_TERM res = create_result_ok(env, state);
+  return res;
+}
+
+UNIFEX_TERM decode_packet(UnifexEnv *env, UnifexNifState *state,
+                          UnifexPayload *in_payload) {
+  char *error = NULL;
+
+  int samples_per_channel = opus_packet_get_nb_samples(
+      in_payload->data, in_payload->size, state->sample_rate);
+  if (samples_per_channel < 0) {
+    error = (char *)opus_strerror(samples_per_channel);
+    goto decode_packet_error;
+  }
+
+  int channels = opus_packet_get_nb_channels(in_payload->data);
+  if (channels < 0) {
+    error = (char *)opus_strerror(channels);
+    goto decode_packet_error;
+  }
+  if (channels != state->channels) {
+    error = "invalid_number_of_channels";
+    goto decode_packet_error;
+  }
+
+  int output_size = samples_per_channel * state->channels * sizeof(opus_int16);
+  UnifexPayload *out_payload =
+      unifex_payload_alloc(env, UNIFEX_PAYLOAD_BINARY, output_size);
+
+  int decoded_samples_per_channel =
+      opus_decode(state->decoder, in_payload->data, in_payload->size,
+                  (opus_int16 *)out_payload->data, out_payload->size, 0);
+  if (decoded_samples_per_channel < 0) {
+    error = (char *)opus_strerror(decoded_samples_per_channel);
+    goto decode_packet_error;
+  }
+  if (decoded_samples_per_channel * state->channels * sizeof(opus_int16) !=
+      output_size) {
```

Comment:
Wouldn't it be enough to test
```suggestion
  if (decoded_samples_per_channel  != samples_per_channel) {
```